File: app/api/schemas/rule_analyzer.py
"""Response schemas for the Rule Analyzer's async investigation API."""
from __future__ import annotations

from datetime import datetime

from pydantic import BaseModel


class InvestigationCreateResponse(BaseModel):
    id: int
    status: str


class InvestigationDetail(BaseModel):
    id: int
    rule_id: int
    customer_id: int
    status: str
    error: str | None = None
    chain_analysis: dict | None = None
    final_report: dict | None = None
    rendered_report: str | None = None
    trace: str | None = None
    tool_calls_made: int | None = None
    created_at: datetime


class InvestigationListItem(BaseModel):
    id: int
    status: str
    tool_calls_made: int | None = None
    created_at: datetime


class InvestigationListResponse(BaseModel):
    items: list[InvestigationListItem]
    total: int
    limit: int
    offset: int

File: frontend/src/lib/api/rule_analyzer.ts
import { apiGet, apiPost } from './client';

export interface RootCause {
	cause: string;
	confidence: 'high' | 'medium' | 'low';
	evidence: string[];
	next_steps: string[];
}

export interface NewRuleOpportunity {
	observation: string;
	evidence: string[];
	suggested_next_step: string;
}

export interface FinalReport {
	detection_intent: string;
	structural_summary: string;
	root_causes: RootCause[];
	overall_recommendation: string;
	additional_findings: NewRuleOpportunity[];
}

export interface ChainAnalysis {
	detection_intent: string;
	preconditions: string[];
	structural_flags: string[];
}

export interface InvestigationCreateResponse {
	id: number;
	status: string;
}

export interface InvestigationDetail {
	id: number;
	rule_id: number;
	customer_id: number;
	status: 'running' | 'completed' | 'failed';
	error: string | null;
	chain_analysis: ChainAnalysis | null;
	final_report: FinalReport | null;
	rendered_report: string | null;
	trace: string | null;
	tool_calls_made: number | null;
	created_at: string;
}

export interface InvestigationListItem {
	id: number;
	status: string;
	tool_calls_made: number | null;
	created_at: string;
}

export interface InvestigationListResponse {
	items: InvestigationListItem[];
	total: number;
	limit: number;
	offset: number;
}

export function startInvestigation(ruleId: number): Promise<InvestigationCreateResponse> {
	return apiPost(`/rule-analyzer/rules/${ruleId}/investigate`, {});
}

export function getInvestigation(investigationId: number): Promise<InvestigationDetail> {
	return apiGet(`/rule-analyzer/investigations/${investigationId}`);
}

export function listInvestigations(
	ruleId: number,
	limit = 20,
	offset = 0
): Promise<InvestigationListResponse> {
	return apiGet(`/rule-analyzer/rules/${ruleId}/investigations`, { limit, offset });
}